libnavigation-core/src/main/java/com/mapbox/navigation/core/telemetry/MapboxNavigationTelemetry.kt

```diff
@@ -217,6 +219,7 @@ internal object MapboxNavigationTelemetry :
             unregisterArrivalObserver(this@MapboxNavigationTelemetry)
         }
         MapboxMetricsReporter.disable()
+        jobControl.scope.cancel()
```

Comment:
Cancel whole scope or children only?

Reply:
Great catch. I believe this should be replaced by `jobControl.job.cancelChildren()` or `jobControl.job.cancel()` right @korshaknn?

Reply:
no, the whole scope. it will be re inited in `init()`
and we must reinit it, because on `MapboxNavigation.onDestroy` we call `ThreadController.cancelAllUICoroutines()`

Reply:
But it seems `MapboxNavigation` is in charge of cancelling the parent scope i.e. shouldn't be `MapboxNavigationTelemetry` responsibility. Are we missing something? 🤔

Reply:
if `telemetry.unregisterListeners` is called we can't use telem until `telemetry.init()` is called.
so in init we create a scope. in unregisterListeners we cancel it.

Reply:
Point here is that the `scope` we're discussing is the "global" Coroutines scope, could cancelling it produce any side effects? 🤔

Reply:
btw we can rename 
`init` -> `start`
`unregisterListeners` -> `stop`

Reply:
```kotlin
fun getMainScopeAndRootJob(): JobControl {
        val parentJob = SupervisorJob(mainRootJob)
        return JobControl(parentJob, CoroutineScope(parentJob + Dispatchers.Main))
}
```

I know we're creating a `CoroutineScope` as part of `getMainScopeAndRootJob` was only making sure that if we cancel it wouldn't affect other coroutines as the `parentJob + Dispatchers.Main` is the same for all of them. If you're confident that won't be a problem, let's leave it as is 👍

Reply:
all scopes  gotten with `getMainScopeAndRootJob `  have the same parent: `parentJob`
when we call `cancelAllUICoroutines` -> `parentJob` cancels it's children, so basically it cancels all UI scopes.
and it cancels our scope too.
and we need to reinit it

Reply:
and it's not a task of MapboxNavigation to handle internal scope of the Telemetry
it should be handled by telemetry itself

Reply:
> I know we're creating a CoroutineScope as part of getMainScopeAndRootJob was only making sure that if we cancel it wouldn't affect other coroutines as the parentJob + Dispatchers.Main is the same for all of them.

the main idea of that method `getMainScopeAndRootJob` is to have a single parent for each UI/non-UI coroutine in the app. It allows us to cancel all of them easily with `parent.cancelChildren`
if we didn't need such cancelation  we would just use `MainScope().launch{}`

Reply:
I think in this particular case is not an issue because we're cancelling all `ThreadController.cancelAllUICoroutines()` right after `MapboxNavigation#onDestroy`

```kotlin
MapboxNavigationTelemetry.unregisterListeners(this@MapboxNavigation)
ThreadController.cancelAllNonUICoroutines()
ThreadController.cancelAllUICoroutines()
```

but i believe we should do `cancelChildren`. What do you think?

Reply:
we reinit it in `init`. so we must `cancel` , not children

Reply:
Why `jobControl.job.cancel()` wouldn't work? Why does it need to be scope? 🤔

Reply:
Isn't the `scope` being cancel as part of `ThreadController.cancelAllUICoroutines()` either way? 🤔

Reply:
@korshaknn you understand this problem better than anyone. So some questions because it seems this problem can only arise when we create multiple MapboxNavigation objects within an app lifecycle (in 1Tap's case, we recreate one when going to the settings screen)

Would like telemetry to only be responsible for the jobs inside telemetry. Confused why `jobControl.job.cancel()` is not what we want.

Could we consider this is an `initialization` issue with `ThreadController`? What if we cancel all old jobs before we initialize a new `ThreadController`
https://github.com/mapbox/mapbox-navigation-android/blob/5592f7154f53de3b2d699f3a7697aad15f1dcd51/libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt#L166

Reply:
> Confused why `jobControl.job.cancel()` is not what we want.

Noting that this is equivalent to `jobControl.scope.cancel()`

```kotlin
public fun CoroutineScope.cancel(cause: CancellationException? = null) {
    val job = coroutineContext[Job] ?: error("Scope cannot be cancelled because it does not have a job: $this")
    job.cancel(cause)
}
```

Reply:
@kmadsen 
`JobControl` is created as 
```kotlin
fun getMainScopeAndRootJob(): JobControl {
        val parentJob = SupervisorJob(mainRootJob)
        return JobControl(parentJob, CoroutineScope(parentJob + Dispatchers.Main))
}
```
when we call `jobControl.scope.cancel()` it `cancels` a parent job of the scope:

```kotlin
public fun CoroutineScope.cancel(cause: CancellationException? = null) {
    val job = coroutineContext[Job] ?: error("Scope cannot be cancelled because it does not have a job: $this")
    job.cancel(cause)
}
```
so,  `jobControl.scope.cancel()` is exactly the same as `jobControl.job.cancel()`
because `jobControl.job == jobControl.scope.coroutineContext[Job]`
we can use both